File: src/Json/Encode.fs
namespace ARCtrl.Json

open Thoth.Json.Core

open Fable.Core

#if FABLE_COMPILER_PYTHON
open Fable.Core.PyInterop
#endif
#if FABLE_COMPILER_JAVASCRIPT
open Fable.Core.JsInterop
#endif

[<RequireQualifiedAccess>]
module Encode = 

    let inline choose (kvs : (string * IEncodable option) list) = 
        kvs
        |> List.choose (fun (k,v) ->
            v
            |> Option.map (fun v -> k,v)
        )       

    /// Try to encode the given object using the given encoder, or return Encode.nil if the object is null
    let tryInclude (name : string) (encoder : 'Value -> IEncodable) (value : 'Value option) = 
        name,
        value
        |> Option.map encoder

    /// Try to encode the given object using the given encoder, or return Encode.nil if the object is null
    let tryIncludeSeq name (encoder : 'Value -> IEncodable) (value : #seq<'Value>) = 
        name,
        if Seq.isEmpty value then None 
        else value |> Seq.map encoder |> Encode.seq |> Some

    let tryIncludeArray name (encoder : 'Value -> IEncodable) (value : 'Value array) = 
        name,
        if Array.isEmpty value then None
        else value |> Array.map encoder |> Encode.array |> Some

    let tryIncludeList name (encoder : 'Value -> IEncodable) (value : 'Value list) = 
        name,
        if List.isEmpty value then None
        else value |> List.map encoder |> Encode.list |> Some

    let tryIncludeListOpt name (encoder : 'Value -> IEncodable) (value : 'Value list option) = 
        name,
        match value with
        | Some(o) -> 
            if List.isEmpty o then None
            else o |> List.map encoder |> Encode.list |> Some
        | _ -> 
            None
        
    let DefaultSpaces = 0

    let defaultSpaces spaces = defaultArg spaces DefaultSpaces

    let dateTime(d : System.DateTime) =         
        d.ToString("O", System.Globalization.CultureInfo.InvariantCulture).Split('+').[0]
        |> Encode.string

    let addPropertyToObject (name : string) (value : Json) (obj : Json) = 
        match obj with
        | Json.Object kvs -> Json.Object (Seq.append kvs [name, value] )
        | _ -> failwith "Expected object"

    let resizeArrayOrSingleton (encoder : 'T -> IEncodable) (values: ResizeArray<'T>) =
        if values.Count = 1 then
            values.[0] |> encoder
        else
            values
            |> Seq.map encoder
            |> Encode.seq

    let dictionary (keyEncoder : Encoder<'key>) (valueEncoder : Encoder<'value>) (values: System.Collections.Generic.IDictionary<'key, 'value>) =
        if values.Count = 0 then
            Encode.nil
        else
            values
            |> Seq.map (fun (KeyValue(k, v)) -> Encode.tuple2 keyEncoder valueEncoder (k,v))
            |> Seq.toArray
            |> Encode.seq

    let intDictionary (valueEncoder : Encoder<'value>) (values: System.Collections.Generic.IDictionary<int, 'value>) =
        if values.Count = 0 then
            Encode.nil
        else
            values
            |> Seq.map (fun (KeyValue(k, v)) ->
                if k > System.Int32.MaxValue || k < System.Int32.MinValue then
                    failwithf "Key %d is out of bounds for Int32" k
                Encode.tuple2 Encode.int valueEncoder (k,v))
            |> Seq.toArray
            |> Encode.seq

File: js/Json/Encode.js
import { defaultArg, map } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { append, map as map_1, isEmpty } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { list as list_1, seq } from "../fable_modules/Thoth.Json.Core.0.4.0/Encode.fs.js";
import { item, map as map_2 } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { map as map_3, isEmpty as isEmpty_1 } from "../fable_modules/fable-library-js.4.22.0/List.js";
import { toString } from "../fable_modules/fable-library-js.4.22.0/Date.js";
import { Json } from "../fable_modules/Thoth.Json.Core.0.4.0/Types.fs.js";

/**
 * Try to encode the given object using the given encoder, or return Encode.nil if the object is null
 */
export function tryInclude(name, encoder, value) {
    return [name, map(encoder, value)];
}

/**
 * Try to encode the given object using the given encoder, or return Encode.nil if the object is null
 */
export function tryIncludeSeq(name, encoder, value) {
    return [name, isEmpty(value) ? undefined : seq(map_1(encoder, value))];
}

export function tryIncludeArray(name, encoder, value) {
    let values;
    return [name, (value.length === 0) ? undefined : ((values = map_2(encoder, value), {
        Encode(helpers) {
            const arg = map_2((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    }))];
}

export function tryIncludeList(name, encoder, value) {
    return [name, isEmpty_1(value) ? undefined : list_1(map_3(encoder, value))];
}

export function tryIncludeListOpt(name, encoder, value) {
    let o;
    return [name, (value != null) ? ((o = value, isEmpty_1(o) ? undefined : list_1(map_3(encoder, o)))) : undefined];
}

export const DefaultSpaces = 0;

export function defaultSpaces(spaces) {
    return defaultArg(spaces, DefaultSpaces);
}

export function dateTime(d) {
    const value = item(0, toString(d, "O", {}).split("+"));
    return {
        Encode(helpers) {
            return helpers.encodeString(value);
        },
    };
}

export function addPropertyToObject(name, value, obj) {
    if (obj.tag === 5) {
        return new Json(5, [append(obj.fields[0], [[name, value]])]);
    }
    else {
        throw new Error("Expected object");
    }
}

